src/cls/queue/cls_queue.cc

```diff
@@ -0,0 +1,1257 @@
+  }
+  CLS_LOG(0, "INFO: cls_create_queue create queue of size %lu", op.head.size);
+  CLS_LOG(1, "INFO: cls_create_queue: Is urgent data present: %d\n", op.head.has_urgent_data);
+
+  // write the head
+  bufferlist bl_head;
+  encode(op.head, bl_head);
+  CLS_LOG(0, "INFO: cls_create_queue writing head of size %u", bl_head.length());
+  ret = cls_cxx_write2(hctx, 0, bl_head.length(), &bl_head, CEPH_OSD_OP_FLAG_FADVISE_WILLNEED);
+  if (ret < 0) {
+    CLS_LOG(0, "ERROR: %s(): cls_cxx_write returned %d", __func__, ret);
+    return ret;
+  }
+  return 0;
+}
+
+static int cls_get_queue_size(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
+{
+  // read the head
+  cls_queue_head head;
+  bufferlist bl_head;
+  int ret = cls_cxx_read2(hctx, 0, QUEUE_HEAD_SIZE, &bl_head, CEPH_OSD_OP_FLAG_FADVISE_WILLNEED);
+  if (ret < 0) {
+    return ret;
+  }
+
+  try {
+    decode(head, bl_head);
+  } catch (buffer::error& err) {
+    CLS_LOG(1, "ERROR: cls_get_queue_size: failed to decode entry\n");
+    return -EINVAL;
+  }
+
+  encode(head.size, *out);
+
+  return 0;
+}
+
+static int cls_enqueue(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
+{
+  CLS_LOG(1, "INFO: cls_enqueue: Read Head of size: %d\n", QUEUE_HEAD_SIZE);
+  // read the head
+  bufferlist bl_head;
+  int ret = cls_cxx_read2(hctx, 0, QUEUE_HEAD_SIZE, &bl_head, CEPH_OSD_OP_FLAG_FADVISE_WILLNEED);
+  if (ret < 0) {
+    return ret;
+  }
+
+  cls_queue_head head;
+  auto iter = bl_head.cbegin();
+  try {
+    decode(head, iter);
+  } catch (buffer::error& err) {
+    CLS_LOG(1, "ERROR: cls_enqueue: failed to decode head\n");
+    return -EINVAL;
+  }
+
+  if (head.front == head.tail && ! head.is_empty) {
+    // return queue full error
+    return -ENOSPC;
+  }
+
+  iter = in->cbegin();
+  cls_enqueue_op op;
+  try {
+    decode(op, iter);
+  } catch (buffer::error& err) {
+    CLS_LOG(1, "ERROR: cls_enqueue: failed to decode input data \n");
+    return -EINVAL;
+  }
+
+  uint64_t total_size = sizeof(op.data.size_data) + op.data.bl_data.length();
+  CLS_LOG(1, "INFO: cls_enqueue(): Total size to be written is %lu and data size is %lu\n", total_size, op.data.size_data);
+
+  bufferlist bl_size;
+  encode(op.data.size_data, bl_size);
+
+  if (head.tail >= head.front) {
+    // check if data can fit in the remaining space in queue
+    if ((head.tail + total_size) <= head.size) {
+      CLS_LOG(1, "INFO: cls_enqueue: Writing data size: offset: %lu, size: %u\n", head.tail, bl_size.length());
+      head.last_entry_offset = head.tail;
+      //write data size at tail offset
+      ret = cls_cxx_write(hctx, head.tail, bl_size.length(), &bl_size);
+      if (ret < 0)
+        return ret;
+
+      head.tail += bl_size.length();
+
+      CLS_LOG(1, "INFO: cls_enqueue: Writing data at offset: %lu\n", head.tail);
+      //write data at tail offset
+      ret = cls_cxx_write2(hctx, head.tail, op.data.bl_data.length(), &(op.data.bl_data), CEPH_OSD_OP_FLAG_FADVISE_SEQUENTIAL);
```

Comment:
ok, the way I did it was resulting in some extra data getting appended to the bufferlist. can you explain what claim_append() does?

Reply:
`a.append(b)` appends all of `bufferlist b`'s buffer segments onto the end of `bufferlist a`. `a.claim_append(b)` is the same, except that it leaves `b` empty. because `claim_append()` transfers ownership of the buffer segments from one list to the other, it doesn't have to modify the reference counts of each segment. but `append()` shares the buffer segments between both lists, so it has to do an atomic increment on each